Another weird failure:

Caused by: org.gradle.api.resources.ResourceException: Could not get resource 'https://artifacts.unidata.ucar.edu/repository/unidata-snapshots/org/bounce/bounce/0.18/bounce-0.18.pom'.

--- a/cdm/core/src/main/java/ucar/nc2/Variable.java
+++ b/cdm/core/src/main/java/ucar/nc2/Variable.java
@@ -7,6 +7,7 @@ package ucar.nc2;
 import com.google.common.base.CharMatcher;
 import com.google.common.base.Splitter;
 import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Iterators;
 import java.util.stream.Collectors;
 import javax.annotation.Nullable;
@@ -1989,18 +1990,18 @@ public class Variable extends CDMNode implements VariableSimpleIF, ProxyReader, 
       return ImmutableList.copyOf(this.dimensions);
     }
 
-    // TODO what about dimString?
-    public List<Dimension> getDimensionsAll() {
-      List<Dimension> dimsAll = new ArrayList<>();
+    // Get all dimension names, including parent structure
+    public ImmutableSet<String> getDimensionsAll() {
+      ImmutableSet.Builder<String> dimsAll = new ImmutableSet.Builder<>();
       addDimensionsAll(dimsAll, this);
-      return dimsAll;
+      return dimsAll.build();
     }
 
-    private void addDimensionsAll(List<Dimension> result, Variable.Builder<?> v) {
-      // if (v.parentStruct != null) TODO
-      // addDimensionsAll(result, v.parentStruct);
-
-      result.addAll(v.dimensions);
+    private void addDimensionsAll(ImmutableSet.Builder<String> result, Variable.Builder<?> v) {
+      if (v.parentStruct != null) {
+        parentStruct.getDimensions().forEach(dim -> result.add(dim.getShortName()));
+      }
+      getDimensionNames().forEach(result::add);
     }
 
     public T setIsScalar() {

--- a/cdm/core/src/main/java/ucar/nc2/dataset/CoordinateSystem.java
+++ b/cdm/core/src/main/java/ucar/nc2/dataset/CoordinateSystem.java
@@ -505,6 +505,14 @@ public class CoordinateSystem {
     return true;
   }
 
+  public static boolean isSubset(Set<String> subset, Set<String> set) {
+    for (String d : subset) {
+      if (!(set.contains(d)))
+        return false;
+    }
+    return true;
+  }
+
   public static Set<Dimension> makeDomain(Iterable<? extends Variable> axes) {
     Set<Dimension> domain = new HashSet<>();
     for (Variable axis : axes) {

--- a/cdm/core/src/main/java/ucar/nc2/dataset/CoordinatesHelper.java
+++ b/cdm/core/src/main/java/ucar/nc2/dataset/CoordinatesHelper.java
@@ -175,7 +175,7 @@ public class CoordinatesHelper {
     public boolean isComplete(CoordinateSystem.Builder<?> cs, VariableDS.Builder<?> vb) {
       Preconditions.checkNotNull(cs);
       Preconditions.checkNotNull(vb);
-      // TODO using strings instead of Dimensions, to avaid exposing mutable Dimension objects.
+      // TODO using strings instead of Dimensions, to avoid exposing mutable Dimension objects.
       // TODO Might reconsider in 6.
       Set<String> varDomain = ImmutableSet.copyOf(vb.getDimensionNames().iterator());
       HashSet<String> csDomain = new HashSet<>();

--- a/cdm/core/src/main/java/ucar/nc2/internal/dataset/CoordSystemBuilder.java
+++ b/cdm/core/src/main/java/ucar/nc2/internal/dataset/CoordSystemBuilder.java
@@ -1,6 +1,7 @@
 package ucar.nc2.internal.dataset;
 
 import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Multimap;
 import java.io.IOException;
 import java.util.ArrayList;
@@ -102,17 +103,18 @@ public class CoordSystemBuilder {
    * @return true if all of the dimensions in the axis also appear in the variable.
    */
   private static boolean isCoordinateAxisForVariable(CoordinateAxis.Builder<?> axis, VariableDS.Builder<?> vb) {
-    List<Dimension> varDims = vb.getDimensionsAll();
-    List<Dimension> axisDims = axis.getDimensionsAll();
+    ImmutableSet<String> varDims = vb.getDimensionsAll();
+    ImmutableSet<String> axisDims = axis.getDimensionsAll();
 
-    // a CHAR variable must really be a STRING, so leave out the last (string length) dimension
-    int checkDims = axisDims.size();
-    if (axis.dataType == DataType.CHAR) {
-      checkDims--;
-    }
+    /*
+     * LOOK a CHAR variable must really be a STRING, so leave out the last (string length) dimension
+     * int checkDims = axisDims.size();
+     * if (axis.dataType == DataType.CHAR) {
+     * checkDims--;
+     * }
+     */
 
-    for (int i = 0; i < checkDims; i++) {
-      Dimension axisDim = axisDims.get(i);
+    for (String axisDim : axisDims) {
       if (!varDims.contains(axisDim)) {
         return false;
       }
@@ -389,13 +391,13 @@ public class CoordSystemBuilder {
       }
 
       // get list of dimensions from '_CoordinateSystemFor' attribute
-      List<Dimension> dimList = new ArrayList<>(6);
+      Set<String> dimList = new HashSet<>();
       StringTokenizer stoker = new StringTokenizer(csVar.coordinateSystemsFor);
       while (stoker.hasMoreTokens()) {
         String dname = stoker.nextToken();
         Optional<Dimension> dimOpt = rootGroup.findDimension(dname);
         if (dimOpt.isPresent()) {
-          dimList.add(dimOpt.get());
+          dimList.add(dimOpt.get().getShortName());
         } else {
           parseInfo.format("***Cant find Dimension %s referenced from CoordSys var= %s%n", dname, csVar);
           userAdvice.format("***Cant find Dimension %s referenced from CoordSys var= %s%n", dname, csVar);
@@ -854,7 +856,7 @@ public class CoordSystemBuilder {
           axis.addAttribute(new Attribute(_Coordinate.ZisPositive, positive));
         }
       }
-      coords.addCoordinateAxis(axis);
+      coords.replaceCoordinateAxis(axis);
       rootGroup.replaceVariable(axis);
       return axis;
     }

--- a/cdm/core/src/main/java/ucar/nc2/internal/dataset/conv/IFPSConvention.java
+++ b/cdm/core/src/main/java/ucar/nc2/internal/dataset/conv/IFPSConvention.java
@@ -1,0 +1,281 @@
+/*
+ * Copyright (c) 1998-2020 John Caron and University Corporation for Atmospheric Research/Unidata
+ * See LICENSE for license information.
+ */
+
+package ucar.nc2.internal.dataset.conv;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import ucar.ma2.Array;
+import ucar.ma2.DataType;
+import ucar.ma2.Index;
+import ucar.ma2.InvalidRangeException;
+import ucar.ma2.Range;
+import ucar.nc2.Attribute;
+import ucar.nc2.Dimension;
+import ucar.nc2.NetcdfFile;
+import ucar.nc2.Variable;
+import ucar.nc2.constants.AxisType;
+import ucar.nc2.constants.CDM;
+import ucar.nc2.constants._Coordinate;
+import ucar.nc2.dataset.CoordinateAxis1D;
+import ucar.nc2.dataset.NetcdfDataset;
+import ucar.nc2.dataset.ProjectionCT;
+import ucar.nc2.dataset.VariableDS;
+import ucar.nc2.dataset.spi.CoordSystemBuilderFactory;
+import ucar.nc2.internal.dataset.CoordSystemBuilder;
+import ucar.nc2.util.CancelTask;
+import ucar.unidata.geoloc.LatLonPointImpl;
+import ucar.unidata.geoloc.Projection;
+import ucar.unidata.geoloc.ProjectionPointImpl;
+import ucar.unidata.geoloc.projection.LambertConformal;
+
+/**
+ * IFPS Convention Allows Local NWS forecast office generated forecast datasets to be brought into IDV.
+ * 
+ * @author Burks
+ */
+public class IFPSConvention extends CoordSystemBuilder {
+  private static final String CONVENTION_NAME = "IFPS";
+
+  public static class Factory implements CoordSystemBuilderFactory {
+    @Override
+    public String getConventionName() {
+      return CONVENTION_NAME;
+    }
+
+    @Override
+    public boolean isMine(NetcdfFile ncfile) {
+      // check that file has a latitude and longitude variable, and that latitude has an attribute called
+      // projectionType
+      boolean geoVarsCheck;
+      Variable v = ncfile.findVariable("latitude");
+      if (null != ncfile.findVariable("longitude") && (null != v)) {
+        geoVarsCheck = (null != v.findAttValueIgnoreCase("projectionType", null));
+      } else {
+        // bail early
+        return false;
+      }
+
+      // check that there is a global attribute called fileFormatVersion, and that it has one
+      // of two known values
+      boolean fileFormatCheck;
+      Attribute ff = ncfile.findGlobalAttributeIgnoreCase("fileFormatVersion");
+      if (ff != null) {
+        String ffValue = ff.getStringValue();
+        // two possible values (as of now)
+        fileFormatCheck = (ffValue.equalsIgnoreCase("20030117") || ffValue.equalsIgnoreCase("20010816"));
+      } else {
+        // bail
+        return false;
+      }
+
+      // both must be true
+      return (geoVarsCheck && fileFormatCheck);
+    }
+
+    @Override
+    public CoordSystemBuilder open(NetcdfDataset.Builder datasetBuilder) {
+      return new IFPSConvention(datasetBuilder);
+    }
+  }
+
+  private Variable.Builder projVar; // use this to get projection info
+
+  private IFPSConvention(NetcdfDataset.Builder datasetBuilder) {
+    super(datasetBuilder);
+    this.conventionName = CONVENTION_NAME;
+  }
+
+  @Override
+  public void augmentDataset(CancelTask cancelTask) throws IOException {
+    if (rootGroup.findVariable("xCoord").isPresent()) {
+      return; // check if its already been done - aggregating enhanced datasets.
+    }
+    parseInfo.format("IFPS augmentDataset %n");
+
+    // Figure out projection info. Assume the same for all variables
+    VariableDS.Builder lonVar = (VariableDS.Builder) rootGroup.findVariable("longitude").get();
+    lonVar.addAttribute(new Attribute(CDM.UNITS, CDM.LON_UNITS));
+    lonVar.addAttribute(new Attribute(_Coordinate.AxisType, AxisType.Lon.toString()));
+    VariableDS.Builder latVar = (VariableDS.Builder) rootGroup.findVariable("latitude").get();
+    latVar.addAttribute(new Attribute(_Coordinate.AxisType, AxisType.Lat.toString()));
+    latVar.addAttribute(new Attribute(CDM.UNITS, CDM.LAT_UNITS));
+
+    projVar = latVar;
+    String projName = projVar.getAttributeContainer().findAttValueIgnoreCase("projectionType", null);
+    if ("LAMBERT_CONFORMAL".equals(projName)) {
+      Projection proj = makeLCProjection();
+      makeXYcoords(proj, latVar, lonVar);
+    }
+
+    // figure out the time coordinate for each data variable
+    // LOOK : always seperate; could try to discover if they are the same
+    for (Variable.Builder ncvar : rootGroup.vbuilders) {
+      // variables that are used but not displayable or have no data have DIM_0, also don't want history, since those
+      // are just how the person edited the grids
+      if ((!ncvar.getFirstDimensionName().equals("DIM_0")) && !ncvar.shortName.endsWith("History")
+          && (ncvar.getRank() > 2) && !ncvar.shortName.startsWith("Tool")) {
+        createTimeCoordinate((VariableDS.Builder) ncvar);
+      } else if (ncvar.shortName.equals("Topo")) {
+        // Deal with Topography variable
+        ncvar.addAttribute(new Attribute(CDM.LONG_NAME, "Topography"));
+        ncvar.addAttribute(new Attribute(CDM.UNITS, "ft"));
+      }
+    }
+  }
+
+  private void createTimeCoordinate(VariableDS.Builder ncVar) {
+    // Time coordinate is stored in the attribute validTimes
+    // One caveat is that the times have two bounds an upper and a lower
+
+    // get the times values
+    Attribute timesAtt = ncVar.getAttributeContainer().findAttribute("validTimes");
+    if (timesAtt == null)
+      return;
+    Array timesArray = timesAtt.getValues();
+
+    // get every other one LOOK this is awkward
+    try {
+      int n = (int) timesArray.getSize();
+      List<Range> list = new ArrayList<>();
+      list.add(new Range(0, n - 1, 2));
+      timesArray = timesArray.section(list);
+    } catch (InvalidRangeException e) {
+      throw new IllegalStateException(e);
+    }
+
+    // make sure it matches the dimension
+    DataType dtype = DataType.getType(timesArray);
+    int nTimesAtt = (int) timesArray.getSize();
+
+    // create a special dimension and coordinate variable
+    Dimension dimTime = ncVar.orgVar.getDimension(0);
+    int nTimesDim = dimTime.getLength();
+    if (nTimesDim != nTimesAtt) {
+      parseInfo.format(" **error ntimes in attribute (%d) doesnt match dimension length (%d) for variable %s%n",
+          nTimesAtt, nTimesDim, ncVar.getFullName());
+      return;
+    }
+
+    // add the dimension
+    String dimName = ncVar.shortName + "_timeCoord";
+    Dimension newDim = new Dimension(dimName, nTimesDim);
+    rootGroup.addDimension(newDim);
+
+    // add the coordinate variable
+    String units = "seconds since 1970-1-1 00:00:00";
+    String desc = "time coordinate for " + ncVar.shortName;
+
+    CoordinateAxis1D.Builder timeCoord = CoordinateAxis1D.builder().setName(dimName).setDataType(dtype)
+        .setDimensionsByName(dimName).setUnits(units).setDesc(desc);
+    timeCoord.setCachedData(timesArray, true);
+    timeCoord.addAttribute(new Attribute(_Coordinate.AxisType, AxisType.Time.toString()));
+    datasetBuilder.replaceCoordinateAxis(rootGroup, timeCoord);
+
+    parseInfo.format(" added coordinate variable %s%n", dimName);
+
+    // now make the original variable use the new dimension
+    List<Dimension> dimsList = ncVar.getDimensions(rootGroup);
+    dimsList.set(0, newDim);
+    ncVar.setDimensions(dimsList);
+
+    // better to explicitly set the coordinate system
+    ncVar.addAttribute(new Attribute(_Coordinate.Axes, dimName + " yCoord xCoord"));
+
+    // fix the attributes
+    Attribute att = ncVar.getAttributeContainer().findAttribute("fillValue");
+    if (att != null)
+      ncVar.addAttribute(new Attribute(CDM.FILL_VALUE, att.getNumericValue()));
+    att = ncVar.getAttributeContainer().findAttribute("descriptiveName");
+    if (null != att)
+      ncVar.addAttribute(new Attribute(CDM.LONG_NAME, att.getStringValue()));
+  }
+
+  protected String getZisPositive() {
+    return "up";
+  }
+
+  private Projection makeLCProjection() {
+    Attribute latLonOrigin = projVar.getAttributeContainer().findAttributeIgnoreCase("latLonOrigin");
+    if (latLonOrigin == null || latLonOrigin.isString())
+      throw new IllegalStateException();
+    double centralLon = latLonOrigin.getNumericValue(0).doubleValue();
+    double centralLat = latLonOrigin.getNumericValue(1).doubleValue();
+
+    double par1 = findAttributeDouble("stdParallelOne");
+    double par2 = findAttributeDouble("stdParallelTwo");
+    LambertConformal lc = new LambertConformal(centralLat, centralLon, par1, par2);
+
+    // make Coordinate Transform Variable
+    ProjectionCT ct = new ProjectionCT("lambertConformalProjection", "FGDC", lc);
+    VariableDS.Builder ctVar = makeCoordinateTransformVariable(ct);
+    ctVar.addAttribute(new Attribute(_Coordinate.Axes, "xCoord yCoord"));
+    rootGroup.addVariable(ctVar);
+
+    return lc;
+  }
+
+  private void makeXYcoords(Projection proj, VariableDS.Builder latVar, VariableDS.Builder lonVar) throws IOException {
+    // brute force
+    Array latData = latVar.orgVar.read();
+    Array lonData = lonVar.orgVar.read();
+
+    Dimension y_dim = latVar.orgVar.getDimension(0);
+    Dimension x_dim = latVar.orgVar.getDimension(1);
+
+    Array xData = Array.factory(DataType.FLOAT, new int[] {x_dim.getLength()});
+    Array yData = Array.factory(DataType.FLOAT, new int[] {y_dim.getLength()});
+
+    LatLonPointImpl latlon = new LatLonPointImpl();
+    ProjectionPointImpl pp = new ProjectionPointImpl();
+
+    Index latlonIndex = latData.getIndex();
+    Index xIndex = xData.getIndex();
+    Index yIndex = yData.getIndex();
+
+    // construct x coord
+    for (int i = 0; i < x_dim.getLength(); i++) {
+      double lat = latData.getDouble(latlonIndex.set1(i));
+      double lon = lonData.getDouble(latlonIndex);
+      latlon.set(lat, lon);
+      proj.latLonToProj(latlon, pp);
+      xData.setDouble(xIndex.set(i), pp.getX());
+    }
+
+    // construct y coord
+    for (int i = 0; i < y_dim.getLength(); i++) {
+      double lat = latData.getDouble(latlonIndex.set0(i));
+      double lon = lonData.getDouble(latlonIndex);
+      latlon.set(lat, lon);
+      proj.latLonToProj(latlon, pp);
+      yData.setDouble(yIndex.set(i), pp.getY());
+    }
+
+    VariableDS.Builder xaxis = VariableDS.builder().setName("xCoord").setDataType(DataType.FLOAT)
+        .setDimensionsByName(x_dim.getShortName()).setUnits("km").setDesc("x on projection");
+    xaxis.addAttribute(new Attribute(CDM.UNITS, "km"));
+    xaxis.addAttribute(new Attribute(CDM.LONG_NAME, "x on projection"));
+    xaxis.addAttribute(new Attribute(_Coordinate.AxisType, "GeoX"));
+
+    VariableDS.Builder yaxis = VariableDS.builder().setName("yCoord").setDataType(DataType.FLOAT)
+        .setDimensionsByName(y_dim.getShortName()).setUnits("km").setDesc("y on projection");
+    yaxis.addAttribute(new Attribute(CDM.UNITS, "km"));
+    yaxis.addAttribute(new Attribute(CDM.LONG_NAME, "y on projection"));
+    yaxis.addAttribute(new Attribute(_Coordinate.AxisType, "GeoY"));
+
+    xaxis.setCachedData(xData, true);
+    yaxis.setCachedData(yData, true);
+
+    rootGroup.addVariable(xaxis);
+    rootGroup.addVariable(yaxis);
+  }
+
+  private double findAttributeDouble(String attname) {
+    Attribute att = projVar.getAttributeContainer().findAttributeIgnoreCase(attname);
+    return (att == null || att.isString()) ? Double.NaN : att.getNumericValue().doubleValue();
+  }
+
+}

--- a/cdm/core/src/main/java/ucar/nc2/internal/dataset/conv/M3IOConvention.java
+++ b/cdm/core/src/main/java/ucar/nc2/internal/dataset/conv/M3IOConvention.java
@@ -1,0 +1,390 @@
+/*
+ * Copyright (c) 1998-2020 John Caron and University Corporation for Atmospheric Research/Unidata
+ * See LICENSE for license information.
+ */
+
+package ucar.nc2.internal.dataset.conv;
+
+import java.io.IOException;
+import java.util.Calendar;
+import java.util.GregorianCalendar;
+import java.util.SimpleTimeZone;
+import java.util.TimeZone;
+import ucar.ma2.ArrayDouble;
+import ucar.ma2.DataType;
+import ucar.nc2.Attribute;
+import ucar.nc2.Dimension;
+import ucar.nc2.NetcdfFile;
+import ucar.nc2.constants.AxisType;
+import ucar.nc2.constants._Coordinate;
+import ucar.nc2.dataset.CoordinateAxis;
+import ucar.nc2.dataset.CoordinateAxis1D;
+import ucar.nc2.dataset.CoordinateTransform;
+import ucar.nc2.dataset.NetcdfDataset;
+import ucar.nc2.dataset.ProjectionCT;
+import ucar.nc2.dataset.VariableDS;
+import ucar.nc2.dataset.spi.CoordSystemBuilderFactory;
+import ucar.nc2.internal.dataset.CoordSystemBuilder;
+import ucar.nc2.util.CancelTask;
+import ucar.unidata.geoloc.ProjectionRect;
+import ucar.unidata.geoloc.projection.AlbersEqualArea;
+import ucar.unidata.geoloc.projection.LambertAzimuthalEqualArea;
+import ucar.unidata.geoloc.projection.LambertConformal;
+import ucar.unidata.geoloc.projection.LatLonProjection;
+import ucar.unidata.geoloc.projection.Mercator;
+import ucar.unidata.geoloc.projection.Stereographic;
+import ucar.unidata.geoloc.projection.TransverseMercator;
+import ucar.unidata.geoloc.projection.UtmProjection;
+
+/**
+ * Models-3/EDSS Input/Output netcf format.
+ * <p/>
+ * The Models-3/EDSS Input/Output Applications Programming Interface (I/O API)
+ * is the standard data access library for both NCSC's EDSS project and EPA's Models-3.
+ * <p/>
+ * 6/24/09: Modified to support multiple projection types of data by Qun He <qunhe@unc.edu> and Alexis Zubrow
+ * <azubrow@unc.edu>
+ * added makePolarStereographicProjection, UTM, and modified latlon
+ * <p/>
+ * 09/2010 plessel.todd@epa.gov add projections 7,8,9,10
+ *
+ * @author plessel.todd@epa.gov
+ * @author caron
+ * @see <a href=
+ *      "http://www.baronams.com/products/ioapi/index.html">http://www.baronams.com/products/ioapi/index.html</a>
+ */
+public class M3IOConvention extends CoordSystemBuilder {
+  private static final String CONVENTION_NAME = "M3IO";
+  private static final double earthRadius = 6370.000; // km
+
+  public static class Factory implements CoordSystemBuilderFactory {
+    @Override
+    public String getConventionName() {
+      return CONVENTION_NAME;
+    }
+
+    @Override
+    public boolean isMine(NetcdfFile ncfile) {
+      return (null != ncfile.findGlobalAttribute("XORIG")) && (null != ncfile.findGlobalAttribute("YORIG"))
+          && (null != ncfile.findGlobalAttribute("XCELL")) && (null != ncfile.findGlobalAttribute("YCELL"))
+          && (null != ncfile.findGlobalAttribute("NCOLS")) && (null != ncfile.findGlobalAttribute("NROWS"));
+    }
+
+    @Override
+    public CoordSystemBuilder open(NetcdfDataset.Builder datasetBuilder) {
+      return new M3IOConvention(datasetBuilder);
+    }
+  }
+
+  /////////////////////////////////////////////////////////////////
+  private CoordinateTransform ct;
+
+  private M3IOConvention(NetcdfDataset.Builder datasetBuilder) {
+    super(datasetBuilder);
+    this.conventionName = CONVENTION_NAME;
+  }
+
+  @Override
+  public void augmentDataset(CancelTask cancelTask) throws IOException {
+    if (rootGroup.findVariable("x").isPresent() || rootGroup.findVariable("lon").isPresent())
+      return; // check if its already been done - aggregating enhanced datasets.
+
+    int projType = rootGroup.getAttributeContainer().findAttributeInteger("GDTYP", 1);
+    boolean isLatLon = (projType == 1);
+    if (isLatLon) {
+      datasetBuilder.replaceCoordinateAxis(rootGroup, makeCoordLLAxis("lon", "COL", "XORIG", "XCELL", "degrees east"));
+      datasetBuilder.replaceCoordinateAxis(rootGroup, makeCoordLLAxis("lat", "ROW", "YORIG", "YCELL", "degrees north"));
+      ct = makeLatLongProjection();
+
+      VariableDS.Builder v = makeCoordinateTransformVariable(ct);
+      rootGroup.addVariable(v);
+      v.addAttribute(new Attribute(_Coordinate.Axes, "lon lat"));
+
+    } else {
+      datasetBuilder.replaceCoordinateAxis(rootGroup, makeCoordAxis("x", "COL", "XORIG", "XCELL", "km"));
+      datasetBuilder.replaceCoordinateAxis(rootGroup, makeCoordAxis("y", "ROW", "YORIG", "YCELL", "km"));
+
+      if (projType == 2)
+        ct = makeLCProjection();
+      else if (projType == 3)
+        ct = makeTMProjection();
+      else if (projType == 4)
+        ct = makeSTProjection();
+      else if (projType == 5)
+        ct = makeUTMProjection();
+      else if (projType == 6) // Was 7. See http://www.baronams.com/products/ioapi/GRIDS.html
+        ct = makePolarStereographicProjection();
+      else if (projType == 7)
+        ct = makeEquitorialMercatorProjection();
+      else if (projType == 8)
+        ct = makeTransverseMercatorProjection();
+      else if (projType == 9)
+        ct = makeAlbersProjection();
+      else if (projType == 10)
+        ct = makeLambertAzimuthalProjection();
+
+      if (ct != null) {
+        VariableDS.Builder v = makeCoordinateTransformVariable(ct);
+        rootGroup.addVariable(v);
+        v.addAttribute(new Attribute(_Coordinate.Axes, "x y"));
+      }
+    }
+
+    makeZCoordAxis("LAY", "VGLVLS", "sigma");
+    makeTimeCoordAxis("TSTEP");
+  }
+
+  @Override
+  protected void makeCoordinateTransforms() {
+    if (ct != null) {
+      VarProcess vp = findVarProcess(ct.getName(), null);
+      vp.isCoordinateTransform = true;
+      vp.ct = CoordinateTransform.builder().setPreBuilt(ct);
+    }
+    super.makeCoordinateTransforms();
+  }
+
+  private CoordinateAxis.Builder makeCoordAxis(String name, String dimName, String startName, String incrName,
+      String unitName) {
+
+    double start = .001 * findAttributeDouble(startName); // km
+    double incr = .001 * findAttributeDouble(incrName); // km
+    start = start + incr / 2.0; // shifting x and y to central
+    CoordinateAxis.Builder v = CoordinateAxis1D.builder().setName(name).setDataType(DataType.DOUBLE)
+        .setDimensionsByName(dimName).setUnits(unitName)
+        .setDesc("synthesized coordinate from " + startName + " " + incrName + " global attributes");
+    v.setAutoGen(start, incr);
+    return v;
+  }
+
+  private CoordinateAxis.Builder makeCoordLLAxis(String name, String dimName, String startName, String incrName,
+      String unitName) {
+    // Makes coordinate axes for Lat/Lon
+    double start = findAttributeDouble(startName); // degrees
+    double incr = findAttributeDouble(incrName); // degrees
+    // The coordinate value should be the cell center.
+    // I recommend also adding a bounds coordinate variable for clarity in the future.
+    start = start + incr / 2.; // shiftin lon and lat to central
+    CoordinateAxis.Builder v = CoordinateAxis1D.builder().setName(name).setDataType(DataType.DOUBLE)
+        .setDimensionsByName(dimName).setUnits(unitName)
+        .setDesc("synthesized coordinate from " + startName + " " + incrName + " global attributes");
+    v.setAutoGen(start, incr);
+    return v;
+  }
+
+  private void makeZCoordAxis(String dimName, String levelsName, String unitName) {
+    Dimension dimz = rootGroup.findDimension(dimName).get();
+    int nz = dimz.getLength();
+    ArrayDouble.D1 dataLev = new ArrayDouble.D1(nz);
+    ArrayDouble.D1 dataLayers = new ArrayDouble.D1(nz + 1);
+
+    // layer values are a numeric global attribute array !!
+    Attribute layers = rootGroup.getAttributeContainer().findAttribute(levelsName);
+    for (int i = 0; i <= nz; i++)
+      dataLayers.set(i, layers.getNumericValue(i).doubleValue());
+
+    for (int i = 0; i < nz; i++) {
+      double midpoint = (dataLayers.get(i) + dataLayers.get(i + 1)) / 2;
+      dataLev.set(i, midpoint);
+    }
+
+    CoordinateAxis.Builder v =
+        CoordinateAxis1D.builder().setName("level").setDataType(DataType.DOUBLE).setDimensionsByName(dimName)
+            .setUnits(unitName).setDesc("synthesized coordinate from " + levelsName + " global attributes");
+    v.setCachedData(dataLev, true);
+    v.addAttribute(new Attribute("positive", "down"));
+    v.addAttribute(new Attribute(_Coordinate.AxisType, AxisType.GeoZ.toString()));
+
+    // layer edges
+    String edge_name = "layer";
+    Dimension lay_edge = new Dimension(edge_name, nz + 1);
+    rootGroup.addDimension(lay_edge);
+    CoordinateAxis.Builder vedge =
+        CoordinateAxis1D.builder().setName(edge_name).setDataType(DataType.DOUBLE).setDimensionsByName(edge_name)
+            .setUnits(unitName).setDesc("synthesized coordinate from " + levelsName + " global attributes");
+    vedge.setCachedData(dataLayers, true);
+    v.setBoundary(edge_name);
+
+    datasetBuilder.replaceCoordinateAxis(rootGroup, v);
+    datasetBuilder.replaceCoordinateAxis(rootGroup, vedge);
+  }
+
+  private void makeTimeCoordAxis(String timeName) {
+    int start_date = findAttributeInt("SDATE");
+    int start_time = findAttributeInt("STIME");
+    int time_step = findAttributeInt("TSTEP");
+
+    int year = start_date / 1000;
+    int doy = start_date % 1000;
+    int hour = start_time / 10000;
+    start_time = start_time % 10000;
+    int min = start_time / 100;
+    int sec = start_time % 100;
+
+    Calendar cal = new GregorianCalendar(new SimpleTimeZone(0, "GMT"));
+    cal.clear();
+    cal.set(Calendar.YEAR, year);
+    cal.set(Calendar.DAY_OF_YEAR, doy);
+    cal.set(Calendar.HOUR_OF_DAY, hour);
+    cal.set(Calendar.MINUTE, min);
+    cal.set(Calendar.SECOND, sec);
+
+    java.text.SimpleDateFormat dateFormatOut = new java.text.SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+    dateFormatOut.setTimeZone(TimeZone.getTimeZone("GMT"));
+
+    String units = "seconds since " + dateFormatOut.format(cal.getTime()) + " UTC";
+
+    // parse the time step
+    hour = time_step / 10000;
+    time_step = time_step % 10000;
+    min = time_step / 100;
+    sec = time_step % 100;
+    time_step = hour * 3600 + min * 60 + sec;
+
+    // create the coord axis
+    CoordinateAxis1D.Builder timeCoord =
+        CoordinateAxis1D.builder().setName("time").setDataType(DataType.INT).setDimensionsByName(timeName)
+            .setUnits(units).setDesc("synthesized time coordinate from SDATE, STIME, STEP global attributes");
+    timeCoord.setAutoGen(0, time_step);
+    timeCoord.addAttribute(new Attribute(_Coordinate.AxisType, AxisType.Time.toString()));
+
+    datasetBuilder.replaceCoordinateAxis(rootGroup, timeCoord);
+  }
+
+  private CoordinateTransform makeLatLongProjection() {
+    // Get lower left and upper right corners of domain in lat/lon
+    double x1 = findAttributeDouble("XORIG");
+    double y1 = findAttributeDouble("YORIG");
+    double x2 = x1 + findAttributeDouble("XCELL") * findAttributeDouble("NCOLS");
+    double y2 = y1 + findAttributeDouble("YCELL") * findAttributeDouble("NROWS");
+
+    LatLonProjection ll = new LatLonProjection("LatitudeLongitudeProjection", new ProjectionRect(x1, y1, x2, y2));
+    return new ProjectionCT("LatitudeLongitudeProjection", "FGDC", ll);
+  }
+
+  private CoordinateTransform makeLCProjection() {
+    double par1 = findAttributeDouble("P_ALP");
+    double par2 = findAttributeDouble("P_BET");
+    double lon0 = findAttributeDouble("XCENT");
+    double lat0 = findAttributeDouble("YCENT");
+
+    LambertConformal lc = new LambertConformal(lat0, lon0, par1, par2, 0.0, 0.0, earthRadius);
+    return new ProjectionCT("LambertConformalProjection", "FGDC", lc);
+  }
+
+  private CoordinateTransform makePolarStereographicProjection() {
+    double lon0 = findAttributeDouble("XCENT");
+    double lat0 = findAttributeDouble("YCENT");
+    double latts = findAttributeDouble("P_BET");
+
+    Stereographic sg = new Stereographic(latts, lat0, lon0, 0.0, 0.0, earthRadius);
+    return new ProjectionCT("PolarStereographic", "FGDC", sg);
+  }
+
+  private CoordinateTransform makeEquitorialMercatorProjection() {
+    double lon0 = findAttributeDouble("XCENT");
+    double par = findAttributeDouble("P_ALP");
+
+    Mercator p = new Mercator(lon0, par, 0.0, 0.0, earthRadius);
+    return new ProjectionCT("EquitorialMercator", "FGDC", p);
+  }
+
+  private CoordinateTransform makeTransverseMercatorProjection() {
+    double lat0 = findAttributeDouble("P_ALP");
+    double tangentLon = findAttributeDouble("P_GAM");
+
+    TransverseMercator p = new TransverseMercator(lat0, tangentLon, 1.0, 0.0, 0.0, earthRadius);
+    return new ProjectionCT("TransverseMercator", "FGDC", p);
+  }
+
+  private CoordinateTransform makeAlbersProjection() {
+    double lat0 = findAttributeDouble("YCENT");
+    double lon0 = findAttributeDouble("XCENT");
+    double par1 = findAttributeDouble("P_ALP");
+    double par2 = findAttributeDouble("P_BET");
+
+    AlbersEqualArea p = new AlbersEqualArea(lat0, lon0, par1, par2, 0.0, 0.0, earthRadius);
+    return new ProjectionCT("Albers", "FGDC", p);
+  }
+
+  private CoordinateTransform makeLambertAzimuthalProjection() {
+    double lat0 = findAttributeDouble("YCENT");
+    double lon0 = findAttributeDouble("XCENT");
+
+    LambertAzimuthalEqualArea p = new LambertAzimuthalEqualArea(lat0, lon0, 0.0, 0.0, earthRadius);
+    return new ProjectionCT("LambertAzimuthal", "FGDC", p);
+  }
+
+  private CoordinateTransform makeSTProjection() {
+    double latt = findAttributeDouble("PROJ_ALPHA");
+    if (Double.isNaN(latt))
+      latt = findAttributeDouble("P_ALP");
+
+    double lont = findAttributeDouble("PROJ_BETA");
+    if (Double.isNaN(lont))
+      lont = findAttributeDouble("P_BET");
+
+    Stereographic st = new Stereographic(latt, lont, 1.0, 0.0, 0.0, earthRadius);
+    return new ProjectionCT("StereographicProjection", "FGDC", st);
+  }
+
+  private CoordinateTransform makeTMProjection() {
+    double lat0 = findAttributeDouble("PROJ_ALPHA");
+    if (Double.isNaN(lat0))
+      lat0 = findAttributeDouble("P_ALP");
+
+    double tangentLon = findAttributeDouble("PROJ_BETA");
+    if (Double.isNaN(tangentLon))
+      tangentLon = findAttributeDouble("P_BET");
+
+    TransverseMercator tm = new TransverseMercator(lat0, tangentLon, 1.0, 0.0, 0.0, earthRadius);
+    return new ProjectionCT("MercatorProjection", "FGDC", tm);
+  }
+
+  private CoordinateTransform makeUTMProjection() {
+    int zone = (int) findAttributeDouble("P_ALP");
+    double ycent = findAttributeDouble("YCENT");
+    boolean isNorth = true;
+    if (ycent < 0)
+      isNorth = false;
+
+    UtmProjection utm = new UtmProjection(zone, isNorth);
+    return new ProjectionCT("UTM", "EPSG", utm);
+  }
+
+  /////////////////////////////////////////////////////////////////////////
+
+  @Override
+  protected AxisType getAxisType(VariableDS.Builder ve) {
+    String vname = ve.shortName;
+
+    if (vname.equalsIgnoreCase("x"))
+      return AxisType.GeoX;
+
+    if (vname.equalsIgnoreCase("y"))
+      return AxisType.GeoY;
+
+    if (vname.equalsIgnoreCase("lat"))
+      return AxisType.Lat;
+
+    if (vname.equalsIgnoreCase("lon"))
+      return AxisType.Lon;
+
+    if (vname.equalsIgnoreCase("time"))
+      return AxisType.Time;
+
+    if (vname.equalsIgnoreCase("level"))
+      return AxisType.GeoZ;
+
+    return null;
+  }
+
+  private double findAttributeDouble(String attname) {
+    return rootGroup.getAttributeContainer().findAttributeDouble(attname, Double.NaN);
+  }
+
+  private int findAttributeInt(String attname) {
+    return rootGroup.getAttributeContainer().findAttributeInteger(attname, 0);
+  }
+
+}

--- a/cdm/core/src/main/java/ucar/nc2/internal/dataset/conv/MADISStation.java
+++ b/cdm/core/src/main/java/ucar/nc2/internal/dataset/conv/MADISStation.java
@@ -1,0 +1,67 @@
+/*
+ * Copyright (c) 1998-2020 John Caron and University Corporation for Atmospheric Research/Unidata
+ * See LICENSE for license information.
+ */
+package ucar.nc2.internal.dataset.conv;
+
+import java.io.IOException;
+import java.util.StringTokenizer;
+import ucar.nc2.Attribute;
+import ucar.nc2.Variable;
+import ucar.nc2.constants.AxisType;
+import ucar.nc2.constants._Coordinate;
+import ucar.nc2.dataset.NetcdfDataset;
+import ucar.nc2.dataset.spi.CoordSystemBuilderFactory;
+import ucar.nc2.internal.dataset.CoordSystemBuilder;
+import ucar.nc2.util.CancelTask;
+
+/** MADIS Station Convention. */
+public class MADISStation extends CoordSystemBuilder {
+  private static final String CONVENTION_NAME = "MADIS_Station_1.0";
+
+  public static class Factory implements CoordSystemBuilderFactory {
+    @Override
+    public String getConventionName() {
+      return CONVENTION_NAME;
+    }
+
+    @Override
+    public CoordSystemBuilder open(NetcdfDataset.Builder datasetBuilder) {
+      return new MADISStation(datasetBuilder);
+    }
+  }
+
+  /////////////////////////////////////////////////////////////////
+
+  private MADISStation(NetcdfDataset.Builder datasetBuilder) {
+    super(datasetBuilder);
+    this.conventionName = CONVENTION_NAME;
+  }
+
+  @Override
+  public void augmentDataset(CancelTask cancelTask) throws IOException {
+    String timeVars = rootGroup.getAttributeContainer().findAttValueIgnoreCase("timeVariables", "");
+    StringTokenizer stoker = new StringTokenizer(timeVars, ", ");
+    while (stoker.hasMoreTokens()) {
+      String vname = stoker.nextToken();
+      rootGroup.findVariable(vname)
+          .ifPresent(v -> v.addAttribute(new Attribute(_Coordinate.AxisType, AxisType.Time.toString())));
+    }
+
+    String locVars = rootGroup.getAttributeContainer().findAttValueIgnoreCase("stationLocationVariables", "");
+    stoker = new StringTokenizer(locVars, ", ");
+    int count = 0;
+    while (stoker.hasMoreTokens()) {
+      String vname = stoker.nextToken();
+      if (rootGroup.findVariable(vname).isPresent()) {
+        Variable.Builder v = rootGroup.findVariable(vname).get();
+        AxisType atype = count == 0 ? AxisType.Lat : count == 1 ? AxisType.Lon : AxisType.Height;
+        v.addAttribute(new Attribute(_Coordinate.AxisType, atype.toString()));
+      } else {
+        parseInfo.format(" cant find time variable %s%n", vname);
+      }
+      count++;
+    }
+  }
+
+}

--- a/cdm/core/src/main/java/ucar/nc2/util/CompareNetcdf2.java
+++ b/cdm/core/src/main/java/ucar/nc2/util/CompareNetcdf2.java
@@ -361,7 +361,7 @@ public class CompareNetcdf2 {
         CoordinateSystem cs2 = copye.getCoordinateSystems().stream().filter(cs -> cs.getName().equals(cs1.getName()))
             .findFirst().orElse(null);
         if (cs2 == null) {
-          f.format("  ** Cant find cs %s in file2 var %s %n", cs1.getName(), org.getShortName());
+          f.format("  ** Cant find CoordinateSystem '%s' in file2 var %s %n", cs1.getName(), org.getShortName());
         } else {
           ok &= compareCoordinateSystem(cs1, cs2, filter);
         }

--- a/cdm/core/src/main/resources/META-INF/services/ucar.nc2.dataset.spi.CoordSystemBuilderFactory
+++ b/cdm/core/src/main/resources/META-INF/services/ucar.nc2.dataset.spi.CoordSystemBuilderFactory
@@ -6,6 +6,9 @@ ucar.nc2.internal.dataset.conv.CoardsConventions$Factory
 ucar.nc2.internal.dataset.conv.CSMConvention$Factory
 ucar.nc2.internal.dataset.conv.GDVConvention$Factory
 ucar.nc2.internal.dataset.conv.GIEFConvention$Factory
+ucar.nc2.internal.dataset.conv.IFPSConvention$Factory
+ucar.nc2.internal.dataset.conv.M3IOConvention$Factory
+ucar.nc2.internal.dataset.conv.MADISStation$Factory
 ucar.nc2.internal.dataset.conv.NUWGConvention$Factory
 ucar.nc2.internal.dataset.conv.WRFConvention$Factory
 ucar.nc2.internal.dataset.conv.ZebraConvention$Factory

--- a/cdm/core/src/test/java/ucar/nc2/dataset/TestCoordSysCompareProblem.java
+++ b/cdm/core/src/test/java/ucar/nc2/dataset/TestCoordSysCompareProblem.java
@@ -25,7 +25,7 @@ public class TestCoordSysCompareProblem {
 
   @Test
   public void compareProblemFile() throws IOException {
-    String filename = TestDir.cdmUnitTestDir + "conventions/nuwg/050231700.lsxmod.nc";
+    String filename = TestDir.cdmUnitTestDir + "conventions/m3io/19L.nc";
     showOrg(filename);
     showNew(filename);
     compare(filename);

--- a/cdm/core/src/test/java/ucar/nc2/dataset/conv/SearchForConventions.java
+++ b/cdm/core/src/test/java/ucar/nc2/dataset/conv/SearchForConventions.java
@@ -29,8 +29,8 @@ import ucar.unidata.util.test.category.NeedsCdmUnitTest;
 @Category(NeedsCdmUnitTest.class)
 @RunWith(Parameterized.class)
 public class SearchForConventions {
-  private static List<String> testDirs = ImmutableList.of(TestDir.cdmUnitTestDir + "/conventions",
-      TestDir.cdmUnitTestDir + "/ft", "E:/data/datasets", "E:/data/work"); // , "F:/bob/data");
+  private static List<String> testDirs =
+      ImmutableList.of(TestDir.cdmUnitTestDir + "/conventions", TestDir.cdmUnitTestDir + "/ft");
   private static Multimap<String, String> convMap = ArrayListMultimap.create();
   private static Multimap<String, String> builderMap = ArrayListMultimap.create();
 
@@ -50,11 +50,11 @@ public class SearchForConventions {
 
   @AfterClass
   public static void showResults() throws IOException {
-    FileWriter out = new FileWriter("C:/Temp/conventions.txt");
+    FileWriter out = new FileWriter("/usr/local/google/home/jlcaron/tmp/conventions.txt");
     showResults(out, convMap);
     out.close();
 
-    out = new FileWriter("C:/Temp/builder.txt");
+    out = new FileWriter("/usr/local/google/home/jlcaron/tmp/builder.txt");
     showResults(out, builderMap);
     out.close();
   }
@@ -80,7 +80,7 @@ public class SearchForConventions {
   }
 
   @Test
-  @Ignore("Not a test - really a utility program")
+  // @Ignore("Not a test - really a utility program")
   public void findConventions() {
     try (NetcdfFile ncfile = NetcdfFiles.open(filename)) {
       System.out.printf("%s%n", filename);

